### PR TITLE
Replace math/rand with crypto/rand for basic-auth password generation

### LINKED ISSUE #273: Swap math/rand with crypto/rand for password generation
Currently basic authentication password use math/rand seeded with the current time. This still has very low entropy and should possibly be replaced with crypto/rand.

There might be other occurrences of math/rand being used in a security relevant context which should also be changed.

### PR DESCRIPTION
**Pull request by plkokanov**
**What this PR does / why we need it**: 
This PR adds a more secure algorithm for generating random numbers used in  basic auth password generation

**Which issue(s) this PR fixes**:
Fixes #273 

**Special notes for your reviewer**: 
The implementation in this PR uses crypto/rand's Int(rand io.Reader, max *big.Int) function to pick a random letter from a predefined string. Currently the password's length is 32 letters so Int() is called 32 times. 
There exist a bit more efficient way to do this by using crypto/rand's Read(b []byte) method to read 24 bytes and then use base64 encoding to create a 32 byte password. However, this will add "-" and "_" characters to the password string which diverges from the current functionality.
Even though the second approach is around 7 times faster (on average) the time it takes to generate passwords in both cases is so minimal that it doesn't really matter that much.

**Release note**: 
NONE

### CODE REVIEW

**Comment by plkokanov** [comment]
@marwinski Could you also have a look (I can't add you as reviewer)

**Comment by dkistner** [comment]
Just to clarify ``math/rand`` is based on time and ``crypto/rand`` reads from ``/dev/urandom``, right?
If this is the case, why not generally use ``crypto/rand``? Then there would be no need to maintain two different ``randomString()`` functions. Ok, it seems a little bit slower, but this shouldn't be any problem.

I have asked @ThormaehlenFred to evaluate if ``/dev/urandom`` is secure enough.

**Comment by ThormaehlenFred** [comment]
I asked Maik Müller, who will return on July 27th.

**Comment by plkokanov** [comment]
@dkistner yes crypto/rand is somewhat slower. It generally depends on the underlying implementation (for unix it's /dev/urandom). Anyway if we only plan on using math/rand for small strings I think it will be ok to swap it entirely to crypto/rand.

**Comment by dkistner** [comment]
I would prefer to use ``crypto/rand`` for the ``GenerateRandomString()`` function  in general, because the Gardener use it to compute also other secrets and for generating some ids. For the ids it wouldn't make a difference, which implementation is used, but for the secrets it would be an improvement with regards to security. The longest random string consist currently of 64 chars, if I remember that correctly.

@mvladev What do you mean?

**Comment by plkokanov** [comment]
Hm, just saw that GenerateRandomStringFromCharset() is also used in [cloud_config.go](https://github.com/gardener/gardener/blob/fe5a126d859e580cf20d2b934ac68b9fd0a873ea/pkg/operation/hybridbotanist/cloud_config.go#L115) where it seems that we also use a very limited set of characters, so this might have to be changed as well.
But otherwise from what I can see in the gardener project GenerateRandomString() is only used for IDs.
To sign certificates and generate RSA keys we already use crypto/rand.

**Comment by dkistner** [comment]
Yes right. Could use also use the ``crypto/rand`` implementation for the mentioned secret in the `cloud_config.go`?

**Comment by plkokanov** [comment]
@dkistner just ran some tests to compare crypto/rand and math/rand on my mac.
math/rand is around 10 milliseconds slower when generating 10000 passwords.  Difference is a lot less with fewer passwords.
I doubt it will be much different on other OSes so I guess we could switch everything to crypto/rand as soon as Maik Müller says it's ok

**Comment by dkistner** [comment]
Ok, ``math/rand`` is even slower then ``crypto/rand``? Then I see no need to continue with ``math/rand``.  It is also not recommend ``math/rand`` in the [Go Code Review comment](https://github.com/golang/go/wiki/CodeReviewComments#crypto-rand). Let's use ``crypto/rand`` in general for all key and ids. Could you please apply the necessary changes?

**Comment by plkokanov** [comment]
@dkistner oops sorry, misstype, meant to say that crypto/rand is slower than math/rand by 10 miliseconds. Still though doesn't seem like much

**Comment by rfranzke** [comment]
I agree with @dkistner that we shouldn't use `math/rand` at all any longer. Please replace everything with `crypto/rand` @plkokanov.

@ThormaehlenFred any update?

**Comment by ThormaehlenFred** [comment]
Mail sent to Maik Mueller who is not on vacation anymore
>From: Thormaehlen, Frederik 
Sent: Dienstag, 31. Juli 2018 10:48
To: Mueller, Maik <maik.mueller@sap.com>
Subject: FW: Is /crypto/rand more secure than /math/rand or secure enough?
Hi Mike,
My colleagues in cc would like to release the PR and asked me again for advice. Could you please have a look on my question below.
Regards Frederik

>From: Thormaehlen, Frederik 
Sent: Dienstag, 24. Juli 2018 15:24
To: Mueller, Maik <maik.mueller@sap.com>
Subject: Is /crypto/rand more secure than /math/rand or secure enough?
Hi Maik,
In this PR
https://github.com/gardener/gardener/pull/284/files
my colleague exchanges the random generator for password to make it more secure.
He exchanges the time based password generation https://golang.org/pkg/math/rand/#Int  with crypto based password generation  https://golang.org/pkg/crypto/rand/#Int which uses /dev/urandom internally.
Can you make an assessment if this is an improvement with respect to generated password security and is the generated password secure enough afterwards from SAP perspective before we release this PR?
Regards
Frederik

**Comment by ThormaehlenFred** [comment]
Answer from Maik Mueller
>I haven’t done a source code review but from high level perspective it is a good idea to use  https://golang.org/pkg/crypto/rand/#Int instead of https://golang.org/pkg/math/rand/#Int.
Regards,  Maik

**Comment by ThormaehlenFred** [comment]
I have nothing to add to Maik's comment who is the expert.

**Comment by rfranzke** [comment]
Sounds good. @plkokanov can you please adapt the PR?

File: pkg/operation/hybridbotanist/cloud_config.go [hunk 1]
```diff
@@ -108,16 +108,23 @@ func (b *HybridBotanist) computeBootstrapToken() (secret *corev1.Secret, err err
...
-		}
-		return b.K8sShootClient.CreateSecret(metav1.NamespaceSystem, secretName, bootstraptokenapi.SecretTypeBootstrapToken, data, true)
+		return b.createBootStrapTokenSecret(tokenID, secretName)
 	}
-
 	return secret, err
 }
+
+func (b *HybridBotanist) createBootStrapTokenSecret(tokenID, secretName string) (secret *corev1.Secret, err error) {
+	bootstrapTokenSecretKey, err := utils.GenerateRandomStringFromCharset(16, "abcdef0123456789")
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/cloud_config.go]: Why did you change the charset (allowed characters were `0123456789abcdefghijklmnopqrstuvwxyz`) ?

File: pkg/utils/random.go [hunk 2]
```diff
@@ -15,27 +15,29 @@
...
 }
 
-// GenerateRandomStringFromCharset generates a random string of the specified length <n>. The set of allowed characters
-// can be specified.
-func GenerateRandomStringFromCharset(n int, allowedCharacters string) string {
-	seed := time.Now().UnixNano()
-	randomSource := rand.NewSource(seed)
-	randomGenerator := rand.New(randomSource)
+// GenerateRandomStringFromCharset generates a cryptographically secure random string of the specified length <n>.
+// The set of allowed characters can be specified. Retyrns error if there was a problem during the random generation.
```
Comment (rfranzke) [inline pkg/utils/random.go]: typo

File: cmd/gardener-controller-manager/app/gardener_controller_manager.go [hunk 3]
```diff
@@ -388,23 +388,37 @@ func makeLeaderElectionConfig(config componentconfig.LeaderElectionConfiguration
...
+		}
 	}
 
 	// If running inside a Kubernetes cluster we will have a service account mount.
 	if ns, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/namespace"); err == nil {
 		gardenerNamespace = string(ns)
 	} else if watchNamespace != nil {
 		gardenerNamespace = *watchNamespace
+	} else {
+		gardenerNamespace = common.GardenNamespace
```
Comment (rfranzke) [inline cmd/gardener-controller-manager/app/gardener_controller_manager.go]: Let's keep it as it was before and initialize `gardenerNamespace` with the `common.GardenNamespace` value right in the beginning (removing complexity of `if-else` construct)

File: pkg/operation/hybridbotanist/cloud_config.go [hunk 4]
```diff
@@ -108,16 +108,23 @@ func (b *HybridBotanist) computeBootstrapToken() (secret *corev1.Secret, err err
...
-			bootstraptokenapi.BootstrapTokenUsageSigningKey:     []byte("true"),
-		}
-		return b.K8sShootClient.CreateSecret(metav1.NamespaceSystem, secretName, bootstraptokenapi.SecretTypeBootstrapToken, data, true)
+		return b.createBootStrapTokenSecret(tokenID, secretName)
 	}
-
 	return secret, err
 }
+
+func (b *HybridBotanist) createBootStrapTokenSecret(tokenID, secretName string) (secret *corev1.Secret, err error) {
```
Comment (rfranzke) [inline pkg/operation/hybridbotanist/cloud_config.go]: is this seperate function really needed?

**Review by rfranzke — COMMENTED** [review]
Thank you @plkokanov.
Found a few things, please take a look.

File: cmd/gardener-controller-manager/app/gardener_controller_manager.go (hunk 3 above)
Comment (plkokanov) [reply cmd/gardener-controller-manager/app/gardener_controller_manager.go]: I'll change it to the way it was

File: pkg/operation/hybridbotanist/cloud_config.go (hunk 4 above)
Comment (plkokanov) [reply pkg/operation/hybridbotanist/cloud_config.go]: Seemed cleaner than embedding the random generation error check in the if statement of the upper function.

File: pkg/utils/random.go (hunk 2 above)
Comment (plkokanov) [reply pkg/utils/random.go]: Will fix it

File: pkg/operation/hybridbotanist/cloud_config.go (hunk 1 above)
Comment (plkokanov) [reply pkg/operation/hybridbotanist/cloud_config.go]: Hm this must've changed recently - I probably didn't rebase/fix merge conflicts properly.

File: cmd/gardener-controller-manager/app/gardener_controller_manager.go [hunk 5]
```diff
@@ -388,16 +388,29 @@ func makeLeaderElectionConfig(config componentconfig.LeaderElectionConfiguration
...
+		gardenerName      string
+		gardenerNamespace string
+		err               error
 	)
 
+	gardenerName, err = os.Hostname()
+	if err != nil {
+		return nil, "", fmt.Errorf("unable to get hostname: %v", err)
+	}
+	gardenerNamespace = common.GardenNamespace
```
Comment (rfranzke) [inline cmd/gardener-controller-manager/app/gardener_controller_manager.go]: merge that with line 393

**Review by rfranzke — APPROVED** [review]
LGTM
